docs/fundamentals/code-analysis/quality-rules/ca2264.md

```diff
@@ -26,7 +26,7 @@ When a value that's known to never be null is passed to `ArgumentNullException.T
 
 ## Rule description
 
-`ArgumentNullException.ThrowIfNull` throws when the passed argument is `null`. Certain constructs like non-nullable structs (except for <xref:System.Nullable%601>), 'nameof()' expressions and 'new' expressions are known to never be null, so `ArgumentNullException.ThrowIfNull` will never throw.
+`ArgumentNullException.ThrowIfNull` throws when the passed argument is `null`. Certain constructs like non-nullable structs (except for <xref:System.Nullable%601>), 'nameof()' expressions, and 'new' expressions are known to never be null, so `ArgumentNullException.ThrowIfNull` will never throw.
```

Comment:
Hmm. Open question here:  What generic constraints would trigger this warning?  Consider:

```csharp
public void M<T>(T arg) where T : struct
{
    ArgumentNullException.ThrowIfNull(...); // I'd expect CA2264 here.
}
```

But, what about:

```
public void M<T>(T arg) where T : notnull
public void M2<T>(T arg) where T : unmanaged
```

I'd expect CA2264 on both methods.
And, finally, I'd expect this wouldn't produce a warning:

```
public void M<T>(T arg) where T : allows ref struct
```

Reply:
@BillWagner What is the action item for me here - do you want me to expand the description, or test these out and then expand the description depending on what I find?

Reply:
@gewarren 

What I wrote above was my guess at what that warning would do. If you've already got a test harness in place, can you test these?  The first three are the only ones I'd expect to trigger this warning. The final one shouldn't.

Reply:
Here's my results:

void M<T>(T arg) where T : struct
{
    ArgumentNullException.ThrowIfNull(arg); // CA2264
}
    
void M1<T>(T arg) where T : notnull
{
    ArgumentNullException.ThrowIfNull(arg); // No violation
}
    
void M2<T>(T arg) where T : unmanaged
{
    ArgumentNullException.ThrowIfNull(arg); // CA2264

}

void M3<T>(T arg) where T : allows ref struct
{
    ArgumentNullException.ThrowIfNull(arg); // CS1503
}

Reply:
Here's what I'd change based on those results:

```suggestion
`ArgumentNullException.ThrowIfNull` throws when the passed argument is `null`. Certain constructs like non-nullable structs (except for <xref:System.Nullable%601>), type parameters known to be non-nullable structs, 'nameof()' expressions, and 'new' expressions are known to never be null, so `ArgumentNullException.ThrowIfNull` will never throw.
```